indexing-hadoop/src/main/java/io/druid/indexer/hadoop/DatasourceRecordReader.java

```diff
@@ -0,0 +1,168 @@
+          @Nullable
+          @Override
+          public String apply(Metric input)
+          {
+            return input.getName();
+          }
+        }
+    );
+
+    complexMetrics = Iterables.filter(
+        spec.getMetrics(), new Predicate<Metric>()
+        {
+          @Override
+          public boolean apply(@Nullable Metric input)
+          {
+            return input.getType() == ValueType.COMPLEX;
+          }
+        }
+    );
+    final DimFilter filter = spec.getFilter();
+
+    String hdfsPath = ((DatasourceInputSplit) split).getPath();
+    logger.info("Reading segment from " + hdfsPath);
+
+    File segmentDir = Files.createTempDir();
+    logger.info("segment dir: " + segmentDir);
+
+    FileSystem fs = FileSystem.get(context.getConfiguration());
+    getSegmentFiles(hdfsPath, segmentDir, fs);
+    logger.info("finished getting segment files");
+
+    QueryableIndex index = IndexIO.loadIndex(segmentDir);
+    StorageAdapter adapter = new QueryableIndexStorageAdapter(index);
+    List<StorageAdapter> adapters = Lists.newArrayList(adapter);
+    rowYielder = new IngestSegmentFirehose(adapters, dimensions, metrics, filter, spec.getInterval(), spec.getGranularity());
+  }
+
+  @Override
+  public boolean nextKeyValue() throws IOException, InterruptedException
+  {
+    if (rowYielder.hasMore()) {
+      currRow = (MapBasedRow)rowYielder.nextRow();
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  @Override
+  public NullWritable getCurrentKey() throws IOException, InterruptedException
+  {
+    return NullWritable.get();
+  }
+
+  @Override
+  public HadoopWritableInputRow getCurrentValue() throws IOException, InterruptedException
+  {
+    Map<String, Object> event = currRow.getEvent();
+    for(Metric m : complexMetrics) {
+      if(event.containsKey(m.getName())) {
+        ComplexMetricSerde serde = ComplexMetrics.getSerdeForType(m.getComplexTypeName());
+        if (serde == null) {
+          throw new ISE("Unknown type[%s]", m.getComplexTypeName());
+        }
+        event.put(m.getName(), serde.getObjectStrategy().toBytes(event.get(m.getName())));
+      }
+    }
+    return new MapBasedHadoopWritableInputRow(new MapBasedInputRow(currRow.getTimestamp(), spec.getDimensions(), event));
+  }
+
+  @Override
+  public float getProgress() throws IOException, InterruptedException
+  {
+    // TODO can we make this more meaningful?
+    return 0;
+  }
+
+  @Override
+  public void close() throws IOException
+  {
+    //TODO: may be delete the copied data even if its in a temp space or maybe make temp directory
+    //also configurable
+  }
+
+  private void getSegmentFiles(String pathStr, File dir, FileSystem fs) throws IOException
+  {
+    if (!dir.exists() && !dir.mkdirs()) {
+      throw new IOException(dir + " does not exist and creation failed");
+    }
+    final Path path = new Path(pathStr);
+    try (FSDataInputStream in = fs.open(path)) {
+      CompressionUtils.unzip(in, dir);
```

Comment:
sure, will check that... haven't tested this code much yet, so some of  these things will show up later. currently, my mind was more into putting in bare minimum to show the implementation details :)